### PR TITLE
Rate limit with express-rate-limit

### PR DESCRIPTION
**Pull request by ChaituVR**

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -2,15 +2,24 @@ import bodyParser from 'body-parser';
...
 export default app => {
   app.use(bodyParser.json({ limit: '20mb' }));
   app.use(bodyParser.urlencoded({ limit: '20mb', extended: false }));
   app.use(frameguard({ action: 'deny' }));
   app.use(cors());
+  app.use(rateLimit({
+    windowMs: 30 * 1000,
+    max: 120,
+    handler: (req, res) => {
+      sendError(res, 'rate limited')
```
Comment (bonustrack) [inline index.ts:20]: Can you add a console.log('Rate limited'); ?

File: index.ts [hunk 2]
```diff
@@ -2,15 +2,24 @@ import bodyParser from 'body-parser';
 import frameguard from 'frameguard';
 import cors from 'cors';
 import { graphqlHTTP } from 'express-graphql';
+import rateLimit from "express-rate-limit";
```
Comment (bonustrack) [inline index.ts:5]:
```suggestion
import rateLimit from 'express-rate-limit';
```

File: index.ts [hunk 3]
```diff
@@ -2,15 +2,25 @@ import bodyParser from 'body-parser';
...
 import { schema, rootValue } from './server/graphql';
 import defaultQuery from './server/graphql/examples';
+import { sendError } from './server/helpers/utils';
 
 export default app => {
   app.use(bodyParser.json({ limit: '20mb' }));
   app.use(bodyParser.urlencoded({ limit: '20mb', extended: false }));
   app.use(frameguard({ action: 'deny' }));
   app.use(cors());
+  app.use(rateLimit({
```
[the remark this answers is not in the document]
Comment (ChaituVR) [reply index.ts:17]: Ahh yes.. sorry didn't notice this in the documentation .. added 👍

**Comment by bonustrack** [comment]
@ChaituVR That should help based on the logs there is quite lot of requests at the same time, and i think it's better to start high so we can see the limitation of the hub, we can adjust if needed.

File: index.ts [hunk 4]
```diff
@@ -2,15 +2,26 @@ import bodyParser from 'body-parser';
...
 import { schema, rootValue } from './server/graphql';
 import defaultQuery from './server/graphql/examples';
+import { sendError } from './server/helpers/utils';
 
 export default app => {
   app.use(bodyParser.json({ limit: '20mb' }));
   app.use(bodyParser.urlencoded({ limit: '20mb', extended: false }));
   app.use(frameguard({ action: 'deny' }));
   app.use(cors());
+  app.set('trust proxy', 1);
```
Comment (sohkai) [inline index.ts:16]: It might be worth putting this behind a env var, since it wouldn't apply to local deployments.

File: index.ts [hunk 5]
```diff
@@ -2,15 +2,26 @@ import bodyParser from 'body-parser';
...
+  app.use(rateLimit({
+    windowMs: 30 * 1000,
+    max: 120,
```
Comment (bonustrack) [inline index.ts:19]:
```suggestion
    max: 12,
```

File: index.ts [hunk 6]
```diff
@@ -2,15 +2,26 @@ import bodyParser from 'body-parser';
...
 
 export default app => {
   app.use(bodyParser.json({ limit: '20mb' }));
   app.use(bodyParser.urlencoded({ limit: '20mb', extended: false }));
   app.use(frameguard({ action: 'deny' }));
   app.use(cors());
+  app.set('trust proxy', 1);
+  app.use(rateLimit({
+    windowMs: 30 * 1000,
+    max: 12,
```
Comment (sohkai) [inline index.ts:19]: I think 12 might be too low for 30s (you might open 10 tabs, for example). What do you think about making the window a bit shorter, e.g. 10s?

I think what we're seeing are very large, sudden bursts of requests from programmatic sources, so the shorter the window the less normal users could be impacted by it.

**Review by sohkai — APPROVED** [review]
🙌 🙌 🙌

File: index.ts [hunk 7]
```diff
@@ -2,15 +2,26 @@ import bodyParser from 'body-parser';
...
+    handler: (req, res) => {
+      console.log('Rate limited');
+      sendError(res, 'rate limited')
+    }
```
Comment (sohkai) [inline index.ts:23]: Ah, we may also want to use the [`keyGenerator`](https://www.npmjs.com/package/express-rate-limit#keygenerator) option to hash the ip.

Although @bonustrack, was there a reason for preferring this to just storing the raw ips in #60?

Comment (bonustrack) [reply index.ts:23]: On #60 the idea was to get a hash of the IP so we can recognize same users in logs without exposing any IPs. But on this console.log we don't expose anything i think that's fine this way.

File: index.ts (hunk 6 above)
Comment (bonustrack) [reply index.ts:19]: I changed this manually just to test a deployment on Heroku. The initial value was 120, i think 12/10 will not be enough, some page in the UI will need multiple request to the API, and we want to allow script to gather data too, we can start with high amount and see how the hub handle it, then lower if needed.

**Review by bonustrack — APPROVED** [review]

**Pull request merged**